NebulaAPI/Extensions/PlayerCollectionExtensions.cs

```diff
@@ -0,0 +1,147 @@
+    {
+        return players.FirstOrDefault(p => p.Connection.Equals(conn));
+    }
+
+    /// <summary>
+    /// Gets a player by their username.
+    /// </summary>
+    /// <param name="players"></param>
+    /// <param name="username"></param>
+    /// <returns></returns>
+    public static INebulaPlayer GetPlayer(
+        this IEnumerable<INebulaPlayer> players, string username)
+    {
+        return players.FirstOrDefault(p => p.Data.Username == username);
+    }
+
+    /// <summary>
+    /// Filters to pending players, with their connection handles as keys.
+    /// </summary>
+    /// <param name="playerConnections"></param>
+    /// <returns></returns>
+    public static IReadOnlyDictionary<INebulaConnection, INebulaPlayer> Pending(
+        this IReadOnlyDictionary<INebulaConnection, INebulaPlayer> playerConnections)
+    {
+        return playerConnections.FilterByConnectionStatus(EConnectionStatus.Pending);
+    }
+
+    /// <summary>
+    /// Filters to syncing players, with their connection handles as keys.
+    /// </summary>
+    /// <param name="playerConnections"></param>
+    /// <returns></returns>
+    public static IReadOnlyDictionary<INebulaConnection, INebulaPlayer> Syncing(
+        this IReadOnlyDictionary<INebulaConnection, INebulaPlayer> playerConnections)
+    {
+        return playerConnections.FilterByConnectionStatus(EConnectionStatus.Syncing);
+    }
+
+    /// <summary>
+    /// Filters to connected players, with their connection handles as keys.
+    /// </summary>
+    /// <param name="playerConnections"></param>
+    /// <returns></returns>
+    public static IReadOnlyDictionary<INebulaConnection, INebulaPlayer> Connected(
+        this IReadOnlyDictionary<INebulaConnection, INebulaPlayer> playerConnections)
+    {
+        return playerConnections.FilterByConnectionStatus(EConnectionStatus.Connected);
+    }
+
+    /// <summary>
+    /// Filters to pending players
+    /// </summary>
+    /// <param name="players"></param>
+    /// <returns></returns>
+    public static IReadOnlyCollection<INebulaPlayer> Pending(
+        this IEnumerable<INebulaPlayer> players)
+    {
+        return players.FilterByConnectionStatus(EConnectionStatus.Pending);
+    }
+
+    /// <summary>
+    /// Filters to syncing players
+    /// </summary>
+    /// <param name="players"></param>
+    /// <returns></returns>
+    public static IReadOnlyCollection<INebulaPlayer> Syncing(
+        this IEnumerable<INebulaPlayer> players)
+    {
+        return players.FilterByConnectionStatus(EConnectionStatus.Syncing);
+    }
+
+    /// <summary>
+    /// Filters to connected players.
+    /// </summary>
+    /// <param name="players"></param>
+    /// <returns></returns>
+    public static IReadOnlyCollection<INebulaPlayer> Connected(
+        this IEnumerable<INebulaPlayer> players)
+    {
+        return players.FilterByConnectionStatus(EConnectionStatus.Connected);
```

Comment:
Since Connected is a frequently used collection, consider caching and updating it only if there are EConnectionStatus changes? Otherwise `.ToList()` will be called multiple times to allocate memory.

Reply:
I plan to bring up some changes to be discussed that may ommit any reason to filter by connection status to begin with - by handling pending and syncing during the connection phase before adding the players to the list.

The performance impact of this is also very negligible considering the maximum amount of players per server can't be very high.

Until that's done or decided to not be done, think we could leave this part as is? Just to keep things simple until then.

Reply:
It's not about the size, it's constantly and rapidly allocation and freeing memory will cause stutter from time to time in game when GC happen. CommonAPI has this kind of issue before.

Reply:
Aha gotcha, I get what you mean now. I went ahead and removed all allocative calls - it should in theory strictly return references now without allocating anything new.

If this is still unsatisfactory I'll add a cache for connection types as well.